### PR TITLE
Make asn1_d2i_read_bio tolerant to BER encoding

### PR DESCRIPTION
**Pull request by bernd-edlinger**
since ASN1_get_object does the same, which uses
asn1_get_length to parse the length field, which
does handle zero-padded BER length encoding
since commit 0e7bda79a1f6 ("Handle BER length encoding.")

Fixes: 35852da1d9e2 ("Add intelligence to asn1_d2i_read_bio for reading entire header without blocking for extra data")

### CODE REVIEW

**Comment by bukka** [comment]
This should probably contain a test in a similar style like #31818 (might be a bit overlapping as it will need similar addition of headers I guess). In any case, it would be nice to add a test here...

File: crypto/asn1/a_d2i_fp.c [hunk 1]
```diff
@@ -199,10 +199,10 @@ int asn1_d2i_read_bio(BIO *in, BUF_MEM **pb)
...
             unsigned int i = *q & 0x7f;
 
-            if (i > sizeof(long)) {
-                ERR_raise(ERR_LIB_ASN1, ASN1_R_TOO_LONG);
-                goto err;
-            }
+            /*
+             * Tolerate BER length encoding which may include leading zeroes,
+             * since ASN1_get_object does also accept BER encoding.
+             */
```
Comment (idrassi) [inline crypto/asn1/a_d2i_fp.c:206]: This is broader than leading zero handling. Memory decode already accepts a long form length count of 0xff when the following 127 length octets are zero. With this change, the BIO path accepts that form too. I can reproduce this with a NULL object: ASN1_get_object and d2i_ASN1_TYPE already accept it, and after this change asn1_d2i_read_bio and ASN1_d2i_bio accept it as well.

That form is not valid BER. X.690 reserves 0xff for future extension. So this change spreads an existing memory decode permissiveness to BIO based decode paths.

File: crypto/asn1/a_d2i_fp.c [hunk 2]
```diff
@@ -199,10 +199,10 @@ int asn1_d2i_read_bio(BIO *in, BUF_MEM **pb)
...
             if (i > diff) {
                 want = q - p + i + 1;
                 continue;
```
Comment (idrassi) [inline crypto/asn1/a_d2i_fp.c:209]: There is also a short read problem here. With a valid zero padded BER NULL using a 9 octet length field, a BIO that returns 2 bytes and then 7 bytes reaches the existing `want == diff` case, calls `BIO_read(..., 0)`, and reports `ASN1_R_NOT_ENOUGH_DATA` while bytes are still pending.

I reproduced this on the PR merge ref. A matrix over valid zero padded counts 1 through 126 had 322 failures with short read BIOs, 315 of them in the newly accepted count range.

**Review by idrassi — CHANGES_REQUESTED** [review]
I understand the compatibility goal but, as noted inline, this change widens more than the objective.

There is also a platform effect on Windows x64 and 32 bit builds, where sizeof(long) is 4. For example, `30 85 00 7f ff ff ff ...` uses 5 raw length octets, so the current BIO pre-reader rejects it early because 5 is greater than sizeof(long). After this change, the header reaches `asn1_get_length`, which skips the leading zero and sees a 4 octet value. That can turn an encoding rejected early today into a very large accepted content length.

File: crypto/asn1/a_d2i_fp.c (hunk 2 above)
Comment (bernd-edlinger) [reply crypto/asn1/a_d2i_fp.c:209]: Well, I don't see how that can happen.
The invariant which should hold here in any case is: `len - off == q - p + diff + 1`
and since `i > diff` and `want = q - p + i + 1`,
the condition `want > diff` should always hold after the continue.
But the
```C
                if ((size_t)i < want)
                    continue;
```
above looks a bit suspicious, as if it could cause an endless loop?

File: crypto/asn1/a_d2i_fp.c (hunk 1 above)
Comment (bernd-edlinger) [reply crypto/asn1/a_d2i_fp.c:206]: At least it everything will be consistent this way.
The final decision whether the encoding is accepted is done by `ASN1_get_object` below.
If such an invalid BER encoding shall not be accepted that will be another PR.

**Comment by bernd-edlinger** [comment]
Update: fixed a possible cause for the ASN1_R_NOT_ENOUGH_DATA reported by [@idrassi](https://github.com/idrassi) (probably an unrelated issue)
And also removed the length check of the tag value, as that may also be zero-padded,
and also since the tag value encoding uses only 7 bits per byte, the length may be 5 bytes even in DER encoding.

File: crypto/asn1/a_d2i_fp.c (hunk 2 above)
Comment (bernd-edlinger) [reply crypto/asn1/a_d2i_fp.c:209]: Aehm, I overlooked the `want -= diff;` at line 134, so no endless loop.

Comment (bernd-edlinger) [reply crypto/asn1/a_d2i_fp.c:209]: Hmm, but the next iteration might indeed execute the `want == diff` condition.
Could it be that you have entered this error condition in this way?

Comment (idrassi) [reply crypto/asn1/a_d2i_fp.c:209]: Yes, that's the path I hit.

For the 9 octet length case, the old code first reads `05 89`, then sets `want` to 11. At the top of the next loop it subtracts the 2 bytes already buffered, so the actual read request is 9. If the BIO returns 7 bytes, the old code keeps `want` as 9 and continues. The next loop then has `diff == 9` and `want == 9`, so `want -= diff` becomes 0 and `BIO_read(..., 0)` returns 0 with input still pending.

I retested the latest head. This part is fixed: the same matrix over zero padded counts 1 through 126 now has 0 failures.

File: crypto/asn1/a_d2i_fp.c (hunk 1 above)
Comment (idrassi) [reply crypto/asn1/a_d2i_fp.c:206]: yes, the final parser call decides acceptance, but in this specific case `ASN1_get_object` doesn't reject the form.

With `05 ff` followed by 127 zero length octets, `ASN1_get_object` accepts the object, reports length 0, and consumes 129 bytes. The current PR makes the BIO path accept that same input as well. That is the part I wanted to call out: this is not only making valid BER with leading zeroes work through BIO, it also spreads acceptance of the reserved `0xff` length count.

**Comment by idrassi** [comment]
@bernd-edlinger On the latest head, the recent fixup removes the high tag length guard in `crypto/asn1/a_d2i_fp.c` around lines 185 to 189. That guard was added in PR #31091 to stop the BIO path from scanning an unbounded high tag field before `ASN1_get_object` parses the buffered header.

I don't think the stated 5 byte DER case explains this removal. The previous check already allowed five high tag octets, because the counter was checked before it was incremented. Also, BER doesn't allow zero padding in the high tag number form. X.690 requires the first subsequent high tag octet not to have all tag bits set to zero.

With the current PR, I can feed a high tag field made of many `0x80` octets followed by `00 00`: the BIO path reads and accepts the whole object. In my test program, a 131072 octet tag case went from quick `ASN1_R_HEADER_TOO_LONG` after 4 reads to successful acceptance after 65537 reads.

The earlier `0xff` length count issue also remains. `ASN1_get_object` accepts `05 ff` followed by 127 zero length octets, and this PR extends that acceptance to the BIO decode path.

**Comment by bernd-edlinger** [comment]
Okay, I see, the 5 bytes limit is not really an issue.  So I reverted that again.
However in order to be BER tolerant the `ASN1_get_object` accepts an unlimited number of leading 0x80 bytes,
so this will remain as one case where `asn1_d2i_read_bio` cannot be BER tolerant, which is unfortunate.
Note that the `ASN1_NOT_ENOUGH_DATA` seems to be a pre-existent unrelated issue:
I assume the loop starts with `diff = 2`, and `want = 6`:
then at the `want -= diff` statement will set `want = 4`.
and the `BIO_read(...,4)` will only return 2 bytes, so `i = 2`.
and now `len += 2`, and after the `continue`
we will have `diff = 4` and `want = 4`
and now `want -= diff` will result in `want = 0`,
therefore the `BIO_read(..., 0) will return 0. and the `ASN_R_NOT_ENOUGH_DATA` case happens.

File: crypto/asn1/a_d2i_fp.c (hunk 1 above)
Comment (bernd-edlinger) [reply crypto/asn1/a_d2i_fp.c:206]: Well, it is easy to make ASN1_get_object reject the 0xFF, maybe that is also acceptable.

**Comment by bernd-edlinger** [comment]
Hmm, you said the BER encoding specified by X.690 does `not` allow a zero-padding of the tag number.
If that is correct, then it would probably be more correct to make `ASN1_get_object` reject such a tag
instead of parsing an unlimited number of leading 0x80 bytes.

File: crypto/asn1/a_d2i_fp.c (hunk 1 above)
Comment (idrassi) [reply crypto/asn1/a_d2i_fp.c:206]: Thanks for the changes. The `0xff` length count part is now addressed.

**Comment by idrassi** [comment]
Thanks for the latest update. The main issues I raised are mostly addressed now.

> Hmm, you said the BER encoding specified by X.690 does `not` allow a zero-padding of the tag number. If that is correct, then it would probably be more correct to make `ASN1_get_object` reject such a tag instead of parsing an unlimited number of leading 0x80 bytes.

Yes, `ASN1_get_object` is the right place to reject this.

There is still one corner case: `1f 00 00` is still accepted as tag 0 with length 0. The first subsequent high tag octet is `00`, so bits 7 to 1 are all zero there too. The current `len == 0` check is inside the `while (*p & 0x80)` loop, so it catches `1f 80 00 00`, but not `1f 00 00`. This means the new check doesn't fully cover the X.690 rule yet.

Also, since the PR now changes both the BIO reader and the core memory parser, I think it is important to have test coverage for the edge cases discussed here.

I think we are very close now. Thanks for iterating on these tricky parser corners!

**Comment by idrassi** [comment]
@bernd-edlinger One detail in the new test data caught my eye. `valid.ber` contains `1f 04 02 00 00`, which encodes tag 4 using the high tag number form.

X.690 section 8.1.2.2 requires tag numbers 0 through 30 to use the single octet identifier form. Section 8.1.2.4 defines the longer form for tag numbers 31 and above. So **this input is not valid BER**.

I also confirmed that the current code accepts `1f 1e 00` as tag 30. The new check covers tag zero and a leading 0x80 octet, but values 1 through 30 remain accepted in this form. As written, the new test records that acceptance as valid BER.

**Comment by bernd-edlinger** [comment]
Okay, thanks.  Now I see.  Well obviously I am not a lawyer ;-)

Actually I am not sure how strict we should make the parsing, as also the latest X.690 spec is not error-free, when I look at `8.1.2.4.3 Figure 4` which illustrates the form of the identifier octets for a type with a tag whose number is greater than 30,
I cannot see how that can possibly be correct, as the last octet and the previous octet have a leading zero in the picture.

What is necessary for consistency between `ASN1_item_d2i` and `ASN1_item_d2i_bio`, is actually only the first
commit 0f62c8d3bac627d93d6471c7594144f538c16a5f.  The following fix-ups were just to make the parsing even more strict,
But it can also create interoperability issues, as previously accepted invalid encodings are now no longer accepted.

Maybe an alternative would be to keep the strictness at a minimum, and also the "invalid.ber" test cases, as an example of invalid but accepted for backward compatibility.  I could instead just adjust the test expectation.

**Review by idrassi — APPROVED** [review]
Thanks for taking the time to work through these tricky cases. Things look good for me now.

Regarding Figure 4 and based on my research, ITU published X.690 Erratum 1 in September 2021. It confirms that the penultimate octet in the original figure was a typo and replaces the figure. The zero bit in the last octet is expected, since it marks the final tag octet. With that correction, the figure agrees with sections 8.1.2.2 and 8.1.2.4.

Of course, the stricter parsing can have a compatibility impact when malformed encodings were accepted in the past. This need to be discussed for stable branches backport but I'm happy with current result on master.

**Comment by bernd-edlinger** [comment]
> Of course, the stricter parsing can have a compatibility impact when malformed encodings were accepted in the past. This need to be discussed for stable branches backport but I'm happy with current result on master.

Okay good point, so I removed the 4.0 and 3.6 labels for now, and will propose a relaxed parsing for the backports.
Thanks!

**Comment by simo5** [comment]
May I ask why on earth we'd want to allow BER encoding in functions that are clearly labeled as handling DER input?

To me it sounds like a recipe for disaster later, where BER ends up being accepted where it is invalid.

**Comment by t8m** [comment]
Yeah, I'd like to see a concrete use case behind this change. The aim should be to tighten things to DER and not to open up for more BER encoding things.

Ideally we should be accepting BER encoding only in clearly defined cases where it is necessary.

File: crypto/asn1/a_d2i_fp.c [hunk 3]
```diff
@@ -164,7 +151,8 @@ int asn1_d2i_read_bio(BIO *in, BUF_MEM **pb)
                     goto err;
                 }
                 len += i;
-                if ((size_t)i < want)
+                want += diff;
+                if (len - off < want)
```
Comment (bernd-edlinger) [inline crypto/asn1/a_d2i_fp.c:155]: Note that this part is a true bug fix. as we discovered:
I assume the loop starts with `diff = 2`, and `want = 6`:
then at the `want -= diff` statement will set `want = 4`.
and the `BIO_read(...,4)` might only return 2 bytes, so `i = 2`.
and now `len += 2`, and after the `continue`
we will have `diff = 4` and `want = 4`
and now `want -= diff` will result in `want = 0`,
therefore the `BIO_read(..., 0)` will return 0. and
now the `ASN_R_NOT_ENOUGH_DATA` case happens.

Comment (bernd-edlinger) [reply crypto/asn1/a_d2i_fp.c:155]: the input that triggers this is `00 84 00 00 00 00`
and by chance the second BIO_read returns only 2 bytes.
